### PR TITLE
S4 architecture

### PR DESCRIPTION
**Pull request by wbenoit26**

### CODE REVIEW

File: projects/train/train/model/supervised.py [hunk 1]
```diff
@@ -20,3 +20,67 @@ def train_step(self, batch: tuple[Tensor, Tensor]) -> Tensor:
...
+        ]  # Unique dicts
+        for hp in hps:
+            params = [
+                p for p in all_parameters if getattr(p, "_optim", None) == hp
+            ]
+            optimizer.add_param_group({"params": params, **hp})
+
+        # Create a lr scheduler
+        scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
+            optimizer, self.trainer.max_epochs
```
Comment (EthanMarx) [inline projects/train/train/model/supervised.py:82]: I think you need to either set this to `self.trainer.estimated_stepping_batches`, or set the `interval` below to `epochs` - double check me on this though

Comment (wbenoit26) [reply projects/train/train/model/supervised.py:82]: True, yeah, missed that

File: libs/architectures/architectures/networks/s4.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        self.s4_layers = nn.ModuleList()
+        self.norms = nn.ModuleList()
+        self.dropouts = nn.ModuleList()
+        for _ in range(n_layers):
+            self.s4_layers.append(
+                S4D(
+                    d_model,
+                    dropout=dropout,
+                    transposed=True,
+                    # lr=min(0.001, args.lr),
```
Comment (EthanMarx) [inline libs/architectures/architectures/networks/s4.py:188]: and then if we do a search I think we will want to mke this an argument somehow

Comment (wbenoit26) [reply libs/architectures/architectures/networks/s4.py:188]: Definitely. Might require a little bit of reconfiguring because we don't normally have the learning rate as an architecture parameter, but should be doable

Comment (EthanMarx) [reply libs/architectures/architectures/networks/s4.py:188]: I think this lr is supposed to be different from the main learning rate, so we could just expose it as a parameter to the architecture. We could also just move the setting of this learning rate from the architecture to the model

Comment (EthanMarx) [reply libs/architectures/architectures/networks/s4.py:188]: Also, would be great to expose the `dt_min`, `dt_max`, and `N` parameters

File: libs/architectures/architectures/networks/s4.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    def __init__(
+        self,
+        d_input,
+        d_output=10,
+        d_model=256,
+        d_state=64,
+        n_layers=4,
+        dropout=0.2,
+        prenorm=False,
+        **kernel_args,
```
Comment (EthanMarx) [inline libs/architectures/architectures/networks/s4.py:174]: @wbenoit26 Was discussing with Christina, can we make the `lr`, `dt_min` and `dt_max` as explicit arguments here instead of hiding them behind `kernel_args`?

File: libs/architectures/architectures/supervised.py [hunk 4]
```diff
@@ -137,3 +137,25 @@ def __init__(
...
+class SupervisedS4Model(S4Model, SupervisedArchitecture):
+    def __init__(
+        self,
+        num_ifos: int,
+        d_output: int = 1,
+        d_model: int = 128,
+        n_layers: int = 4,
+        dropout: float = 0.1,
+        prenorm: bool = True,
+        **kernel_args,
```
Comment (EthanMarx) [inline libs/architectures/architectures/supervised.py:151]: and then expose `dt_min` , `dt_max`, `lr` here as well

File: libs/architectures/architectures/networks/s4.py (hunk 3 above)
Comment (wbenoit26) [reply libs/architectures/architectures/networks/s4.py:174]: Sure, I'll update this

File: libs/architectures/architectures/networks/s4.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        self.dropout = dropout_fn(dropout) if dropout > 0.0 else nn.Identity()
+
+        # position-wise output transform to mix features
+        self.output_linear = nn.Sequential(
+            nn.Conv1d(self.h, 2 * self.h, kernel_size=1),
+            nn.GLU(dim=-2),
+        )
+
+    def forward(
+        self, u, **kwargs
```
Comment (EthanMarx) [inline libs/architectures/architectures/networks/s4.py:143]: Can we remove `**kwargs` if its not used anywhere?

**Review by EthanMarx — APPROVED** [review]
